helpers/apolloClient-server.ts

```diff
@@ -0,0 +1,63 @@
+import {
+  ApolloClient,
+  InMemoryCache,
+  NormalizedCacheObject
+} from '@apollo/client'
+import merge from 'deepmerge'
+import isEqual from 'lodash/isEqual'
+import { SchemaLink } from '@apollo/client/link/schema'
+import { schema } from '../graphql/schema'
+
+let apolloClient: ApolloClient<NormalizedCacheObject>
+const cache = new InMemoryCache()
+
+function createIsomorphLink() {
+  if (typeof window === 'undefined') {
+    return new SchemaLink({ schema })
+  } else {
+    const { HttpLink } = require('@apollo/client/link/http')
+    return new HttpLink({
+      uri: '/api/graphql',
+      credentials: 'same-origin'
+    })
+  }
+}
+
+function createApolloClient() {
+  return new ApolloClient({
+    ssrMode: typeof window === 'undefined',
+    link: createIsomorphLink(),
+    cache
+  })
+}
+
+export function initializeApollo(
+  initialState: NormalizedCacheObject | null = null
+) {
+  const _apolloClient = apolloClient || createApolloClient()
+
+  // If your page has Next.js data fetching methods that use Apollo Client, the initial state
+  // gets hydrated here
+  if (initialState) {
+    // Get existing cache, loaded during client side data fetching
+    const existingCache = _apolloClient.extract()
+
+    // Merge the existing cache into data passed from getStaticProps/getServerSideProps
+    const data = merge(initialState, existingCache, {
+      // combine arrays using object equality (like in sets)
+      arrayMerge: (destinationArray, sourceArray) => [
+        ...sourceArray,
+        ...destinationArray.filter(d => sourceArray.every(s => !isEqual(d, s)))
+      ]
+    })
+
+    // Restore the cache with the merged data
+    _apolloClient.cache.restore(data)
+  }
+  // For SSG and SSR always create a new Apollo Client
+  if (typeof window === 'undefined') return _apolloClient
```

Comment:
we don't need this `if` check because it's server side apolloClient so window should always be `undefined`